### PR TITLE
Remove `Local2Int` function for performance and stability

### PR DESCRIPTION
**Pull request by dkcumming**
`Local2Int` is regularly a side condition on a rule, and this may be causing problems for the legacy and booster backends. This change makes `<localDecl>` index cell `<index>` take a `Local` instead of an `Int`, which removes the need to for `Local2Int` altogether.

### Proof that was getting stuck on `evalLocal`
@Baltoli Success!
```
    claim [failing]: 
    <k> #assert( FN_KEY, _1, _KIND) => .K ...</k>
    <mir>
      <data>
        <functions>
            <function>
            <fnKey> FN_KEY </fnKey>
            <localDecls>
              <localDecl>
              <index> _1 </index>
                <value> true:Bool </value>
                ...
              </localDecl>
              <localDecl>
              <index> _0 </index>
                <value> 8 </value>
                ...
              </localDecl>
              ...
            </localDecls>
            ...
            </function>
        </functions>
      </data>
      ...
    </mir>
```
```
daniel@daniel-MS-7E06 kmir$ poetry run kmir prove src/tests/integration/test-data/prove-rs/lemmas.k --save-directory kcfgs
PROOF PASSED: LEMMAS.failing
```

### Evaluations of performance with booster
@geo2a Doesn't seem to change the performance. However this is good, remember the proposed alternative was #292 which created massive slowdown. This doesn't solve the problem, but it doesn't exacerbate it. At least this change should simplify things.

##### With booster
```
daniel@daniel-MS-7E06 kmir$ time poetry run kmir prove --use-booster src/tests/integration/test-data/prove-rs/simple-spec.k
//-snip-
PROOF PASSED: SIMPLE-SPEC.statements.1
PROOF PASSED: SIMPLE-SPEC.statement.1
PROOF PASSED: SIMPLE-SPEC.statement.3
PROOF PASSED: SIMPLE-SPEC.statement.4
PROOF PASSED: SIMPLE-SPEC.statement.2
PROOF PASSED: SIMPLE-SPEC.statements.2

real    3m15.456s
user    3m18.684s
sys     0m8.333s
```
##### Without booster
```
daniel@daniel-MS-7E06 kmir$ time poetry run kmir prove src/tests/integration/test-data/prove-rs/simple-spec.k 
PROOF PASSED: SIMPLE-SPEC.statements.1
PROOF PASSED: SIMPLE-SPEC.statement.1
PROOF PASSED: SIMPLE-SPEC.statement.3
PROOF PASSED: SIMPLE-SPEC.statement.4
PROOF PASSED: SIMPLE-SPEC.statement.2
PROOF PASSED: SIMPLE-SPEC.statements.2

real    1m2.728s
user    1m11.782s
sys     0m3.738s
```

##### With booster
```
daniel@daniel-MS-7E06 kmir$ time poetry run kmir prove --use-booster src/tests/integration/test-data/prove-rs/simple-spec.k
//-snip-
PROOF PASSED: SIMPLE-SPEC.statement.3
PROOF PASSED: SIMPLE-SPEC.statement.2
PROOF PASSED: SIMPLE-SPEC.statements.1
PROOF PASSED: SIMPLE-SPEC.statements.2
PROOF PASSED: SIMPLE-SPEC.statement.1
PROOF PASSED: SIMPLE-SPEC.statement.4

real    3m11.596s
user    3m14.730s
sys     0m8.368s
```
##### Without booster
```
daniel@daniel-MS-7E06 kmir$ time poetry run kmir prove src/tests/integration/test-data/prove-rs/simple-spec.k 
PROOF PASSED: SIMPLE-SPEC.statement.3
PROOF PASSED: SIMPLE-SPEC.statement.2
PROOF PASSED: SIMPLE-SPEC.statements.1
PROOF PASSED: SIMPLE-SPEC.statements.2
PROOF PASSED: SIMPLE-SPEC.statement.1
PROOF PASSED: SIMPLE-SPEC.statement.4

real    1m0.706s
user    1m9.662s
sys     0m3.690s
```

### CODE REVIEW

File: kmir/src/tests/integration/test-data/prove-rs/simple-spec.k [hunk 1]
```diff
@@ -63,29 +61,28 @@ module SIMPLE-SPEC
...
               <localDecl>
-                <index> INDEX2 </index>
+                <index> PLACE2 </index>
                 <value> _ => 9 </value>
                 ...
               </localDecl>
               ...
           </function>
-    requires INDEX1 ==Int Local2Int(PLACE1) andBool INDEX2 ==Int Local2Int(PLACE2) andBool INDEX1 =/=Int INDEX2
+    requires PLACE1 =/=K PLACE2
```
Comment (dkcumming) [inline kmir/src/tests/integration/test-data/prove-rs/simple-spec.k:75]: This change will mean that we have more `==K` and `=/=K`

Comment (yanliu18) [reply kmir/src/tests/integration/test-data/prove-rs/simple-spec.k:75]: It looks good to me.
Furthermore, will it be a good idea to extend the builtin types with `Local`, which is to implement a `==Local`, `=/=Local` and other operations.

It doesnot seem immediately necessary though. What do you think?

**Comment by yanliu18** [comment]
@dkcumming thank you. I like the way you are collecting the timing information of each prove run.
This could give us a better idea of what changes improve the performance.

Would you please keep track of these stas somewhere (maybe a file or an issue)? Ideally for each `prove` related PR.
with `prove_time` as a file name or title, with data format as follows:

| commits | use booster| real | user | sys |
|----------|------------|-----|-----|-----|
|[master@41c2a6a](https://github.com/runtimeverification/mir-semantics/commit/41c2a6ab2bf36f6696875334af8891b33dc3e67b) | y | 3m11.596s | 3m14.730s | 0m3.690s |

**Review by yanliu18 — APPROVED** [review]

File: kmir/src/tests/integration/test-data/prove-rs/simple-spec.k (hunk 1 above)
Comment (dkcumming) [reply kmir/src/tests/integration/test-data/prove-rs/simple-spec.k:75]: I was having this thought as well. I think for now we can leave it as `==K`, but there may be a point in the future where a change is required.

**Comment by dkcumming** [comment]
> @dkcumming thank you. I like the way you are collecting the timing information of each prove run. This could give us a better idea of what changes improve the performance.
> 
> Would you please keep track of these stas somewhere (maybe a file or an issue)? Ideally for each `prove` related PR. with `prove_time` as a file name or title, with data format as follows:
> 
> commits	use booster	real	user	sys
> [master@41c2a6a](https://github.com/runtimeverification/mir-semantics/commit/41c2a6ab2bf36f6696875334af8891b33dc3e67b)	y	3m11.596s	3m14.730s	0m3.690s

This is a great idea, I will create an issue tracking this!

**Pull request merged**